### PR TITLE
adding aws testing infra

### PR DESCRIPTION
**Pull request by RinkiyaKeDad**
Signed-off-by: RinkiyaKeDad <arshsharma461@gmail.com>

Making a WIP PR so that reviewing errors is easier 🙂

### CODE REVIEW

**Comment by RinkiyaKeDad** [comment]
@jakexks here's the complete error message:

```
Error: error deleting subnet (subnet-010a94bd448bec453): timeout while waiting for state to become 'destroyed' (last state: 'pending', timeout: 20m0s)



Error: Kubernetes cluster unreachable: stat ./kubeconfig_cert-manager-cluster: no such file or directory



Error: Error waiting for internet gateway (igw-0063ea481a3796a7e) to detach: timeout while waiting for state to become 'detached' (last state: 'detaching', timeout: 15m0s)



Error: error deleting subnet (subnet-0be126a106b762223): timeout while waiting for state to become 'destroyed' (last state: 'pending', timeout: 20m0s)



Error: Kubernetes cluster unreachable: stat ./kubeconfig_cert-manager-cluster: no such file or directory



Error: Error deleting VPC: DependencyViolation: The vpc 'vpc-088e1c7a864ce269f' has dependencies and cannot be deleted.
        status code: 400, request id: 71fe787d-c7e3-431f-b13c-bafac3708c37
```

File: aws/main.tf [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      instance_type = var.nodes_type
+    }
+  }
+
+  write_kubeconfig = var.write_kubeconfig
+}
+
+provider "helm" {
+  kubernetes {
+    config_path = "${path.module}/kubeconfig_cert-manager-cluster" ## check this
```
Comment (RinkiyaKeDad) [inline aws/main.tf:73]: I did use `${path.module}` here but even then it fails midway saying the file doesn't exist but when I run `terraform apply` again then it works.

File: aws/main.tf [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  }
```
Comment (jakexks) [inline aws/main.tf:74]:
```suggestion
  kubernetes {
    host                   = data.aws_eks_cluster.cluster.endpoint
    cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster.certificate_authority.0.data)
    exec {
      api_version = "client.authentication.k8s.io/v1alpha1"
      args        = ["eks", "get-token", "--cluster-name", module.eks.cluster_id]
      command     = "aws"
    }
  }

```

Terraform can construct a dependency graph if you directly reference other terraform resources. The reason you're getting an error some of the time is you're referencing a direct path to a config file that may or may not be there at the time terraform evaluates it

File: aws/main.tf [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  name       = "ingress-nginx"
+  repository = "https://kubernetes.github.io/ingress-nginx"
+  chart      = "ingress-nginx"
+  namespace  = "default"
+}
+
+data "kubernetes_service" "ingress_service" {
+  metadata {
+    name = "ingress-nginx-controller"
+  }
```
Comment (jakexks) [inline aws/main.tf:100]:
```suggestion
  }
  
  depends_on = [
    helm_release.ingress_nginx
  ]
```

Similar here, this service doesn't exist until after it has been created

**Review by jakexks — CHANGES_REQUESTED** [review]

**Comment by RinkiyaKeDad** [comment]
Thanks for reviewing @jakexks! I added the changes you suggested but I'm still not able to figure out how to get out of the current state I'm stuck in. After these changes

When I run `terraform apply` I get:
```
Error: Kubernetes cluster unreachable: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable



Error: Kubernetes cluster unreachable: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
```

And if I run `terraform destroy` I get:
```
Error: Kubernetes cluster unreachable: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable



Error: Error waiting for internet gateway (igw-0063ea481a3796a7e) to detach: timeout while waiting for state to become 'detached' (last state: 'detaching', timeout: 15m0s)



Error: error deleting subnet (subnet-010a94bd448bec453): timeout while waiting for state to become 'destroyed' (last state: 'pending', timeout: 20m0s)



Error: Error deleting VPC: DependencyViolation: The vpc 'vpc-088e1c7a864ce269f' has dependencies and cannot be deleted.
        status code: 400, request id: cda89d89-3ec8-4042-b0a3-78a6f6e01330



Error: Kubernetes cluster unreachable: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
```


Is there a way to reset everything and start from scratch with these new changes? 😅

**Comment by RinkiyaKeDad** [comment]
Update: So I deleted all the files terraform generates and just let the `main.tf` and `variables.tf` be. This time `terraform apply` worked in one go without stopping in mid and gave the load balancer hostname as the output. Even `terraform destroy` worked without errors. So YAAAAY!

**Comment by RinkiyaKeDad** [comment]
@jakexks I forgot that you had mentioned that we don't even need the load balancer hostname as output now since we use the route 53 zone. This means the error during `terraform apply` doesn't occur now. With that out of the way, the PR is ready please review when free :)

**Review by jakexks — APPROVED** [review]
/lgtm

**Pull request merged**